### DISCUSSION TITLE
4.2.7 board Wiki - For dying 4.2.2 boards

### DISCUSSION THREAD

**Discussion by GalvanicGlaze** [Jyers/Marlin#35] [Ideas]
I don't know how hard it is to change to a 4.2.7 in terms for compiling firmware but seeing as how lots of ender 3 v2 4.2.2 boards are dying and doing layer shifting it would be great to have a tutorial on how to flash to an updated 4.2.7 board

**Comment by amandamana** [comment]
Yes. I missed this firmware after switching from a 4.2.2 to a 4.2.7 to fix layer shifts.

[Jyvers wrote](https://www.reddit.com/r/ender3v2/comments/l4vbuj/the_new_and_improved_v2_menu/gm457jh/) earlier: 
> The only reason I don't compile for that board is because I have to recompile for every configuration after each bugfix and adding 4.2.7 versions would effectively double the amount of firmware I have to compile, which is quite time consuming. As far as I can tell from my research, the only change that needs to be made for the 4.2.7 board is changing the TMC2208 to the TMC2225 in the configuration files. From there you can compile your own firmware, there are plenty of tutorials online for that. 

I've installed the software needed to compile Marlin, and I believe the **only change** needed is in _Configuration.h_ to change:
`#define MOTHERBOARD BOARD_CREALITY_V4` to `#define MOTHERBOARD BOARD_CREALITY_V427`

That change should account for the different pin assignments between the drivers, and I don't think we need to change "TMC2208" in view of the source code comment:
` * Use TMC2208/TMC2208_STANDALONE for TMC2225 drivers`

It would be good to get a confirmation from an experienced individual before I flash my build.

**Reply by brett0712** [reply]
That is all that is needed - just change MOTHERBOARD BOARD_CREALITY_V4 to MOTHERBOARD BOARD_CREALITY_V427. You are correct in that you don't need to change the stepper driver settings as TMC2208_STANDALONE covers the TMC2225 drivers used on the 4.2.7 board.

I did the very same thing as I replaced my 4.2.2 boards with 4.2.7s due to layer shift issues. I am running BLTouchs as well (Creality kit with single cable for BLTouch port) and everything worked flawlessly. My E3V2s have been running 24/7 for about 5 days with this firmware (previous version I compiled about a week ago) without a single issue. 

I really look forward to giving this new update a try.

**Reply by Jyers** [reply]
Thanks for this note, I'll add a bit to the wiki page on firmware flashing to explain these changes!

**Comment by jisr007** [comment]
I also compiled your firmware just by changing the motherboard, I also have a 4.2.7 and I have been printing for about 20 hours without problems of any kind.

**Comment by amandamana** [comment]
I'm considering how much of perceived barrier there is for novice users to compile Marlin for the first time, and how you might best spend your precompiling time to serve the most users. Right now you're publishing 7 builds, which must take about 30 minutes if not automated. Some of these builds target more advanced users, who I expect are generally more capable of compiling Marlin themselves, and probably do that anyways to customize their bed sizes etc. Some of the changes are also quite trivial, e.g., a single line change between normal and high probe speeds. You haven't published configs for 5x5 meshes (would you please?), but I imagine they're not much different from 3x3, in other words: easy for builders to tweak themselves.

What do you think about dropping some of your precompiled builds in favor of adding 4.2.7 support? Perhaps just provide one probe speed at a single mesh size in the future?

Also consider that you don't necessarily need to double your work filling out the whole build matrix with 4.2.7 flavours. I think the RunoutSensor option is already a one-off build, with that configuration absent from others.

Right now the killer feature of your Extensible UI is manual mesh, which isn't possible with stock firmware. Very few beds are perfectly flat and E2V2 users with warped beds need to add hardware to manage this, such as a probe, and when they do that, Creality already provides firmware to support them. So if you only add one 4.2.7 build, I think Manual Mesh would be the best choice.

Perhaps you could drop the Default build entirely. Correct me if I'm wrong, but there isn't any disadvantage to installing the Manual Mesh build instead of Default. That build functions the same as Default if the user doesn't activate the Mesh function, and then it's there if they want it.

Staying with 7 builds, my suggestion would be:
```
BL Touch 3x3
BL Touch 5x5
BL Touch 5x5 HS
Default NoProbe
Manual Mesh 3x3
Manual Mesh 3x3 RunoutSensor
Manual Mesh 3x3 v427 board
```
And to cut down the BL Touch builds to only one, and/or drop Default, if you want to reduce your work or make room for more important builds, such as:
```
BL Touch 3x3
BL Touch 5x5 HS
Manual Mesh 3x3
Manual Mesh 3x3 RunoutSensor
Manual Mesh 3x3 v427 board
Manual Mesh 5x5
Manual Mesh 5x5 v427 board
```
I'm curious what you think. Perhaps you could even conduct a vote. I think we can agree that v4.2.7 boards are only going to become more common going forward.

Of course the best solution would be to unify some of the builds by converting some of the compile-time options into UI settings stored to the EEPROM. I don't know how difficult (or possible) that might be given the present design of Marlin. Perhaps it's possible to enable (un-disable) a RunoutSensor or change the Mesh size after compile time?

**Reply by Jyers** [reply]
Thank you so much for this! I really appreciate how thorough you've been. I think my primary goal might be to try to set up some kind of automated way to compile configurations, that way I can compile as many as I want without much added time. If I can't figure out an effective way of doing this I will definitely think on some of your suggestions. I do think I should probably cut down on BL Touch configs and maybe even remove the default as you mentioned. Also adding v427 boards is definitely a must moving forward. The only thing I probably won't do is the addition of a runout sensor config. As properly configuring a filament runout sensor requires tuning the config to your printers spool to extruder distance and such, so that doesn't really make sense as a precompiled firmware. Adding config-modifications to the advanced menu on screen has been something I've wanted to do for a while but haven't really had the time to look into much yet. I would especially like a grid size option, but that may require UBL to function properly... By the way the reason there are no 5x5 configs is because its only one setting change is is a pretty easy to find in the leveling section of configuration.h, so I figured I'd leave it to each person to decide their grid size and set it themselves.

**Reply by Sennevds** [reply]
Maybe GitHub actions could be used: https://docs.platformio.org/en/latest/integration/ci/github-actions.html the only thing needed is to create multiple actions which copies the right configuration.h and configuration_adv.h file and then pushes the bin file to the right directory

**Reply by luketot** [reply]
I compiled my own, but in an effort to ease your workload jyers, I think amandamana's suggestion of enabling the runout sensor makes sense. I thought I saw a define for whether the sensor is enabled by default, and then users with the sensor simply have to send M412 S1 to enable it.

Similarly, any marlin features that don't interfere when enabled should be enabled. I think they're disabled to save memory for 8-bit boards. Larger buffer, host action support, emergency parser, meatpack, faster baud rate, etc.

**Reply by Jyers** [reply]
I definitely agree, filament runout specifically will probably be enabled with the merge of #97. If there are other configuration changes you think would be beneficial, I would recommend a PR with the changes as config additions aren't very high on my priority list right now.

**Reply by t3chguy** [reply]
The UI doesn't let you configure whether your filament runout sensor is active-high or active-low nor pull-ups which depend on the runout sensor + wiring you are using :/

**Comment by bakaneko59** [comment]
I don't know squat about this, but I really like amandamana's suggestion of putting various configurations into the user's hands via a "preferences" menu or something. Theoretically (if there's enough program space,) you could end up with single build with user being in control of whether or not they have a bltouch, whether the have a 4.2.2 vs 4.2.7,what mesh size they want etc. The capability to save custom values obviously exists since we can change our defaults Temps for PLA etc.

**Reply by Jyers** [reply]
While in an ideal world this would be great, such an extreme case isn't possible. A large amount of the program is decided at compile time based on certain configurations. So such settings as the presence of a BL Touch, the kind of leveling, or the inclusion of a certain feature are impossible to implement in the menu as the code for them is add/removed at compile time and therefore can not be used if not initially configured that way.

**Reply by bakaneko59** [reply]
XWell, like I said - I don't know squat. It's too bad though, cuz you would only have to maintain  one master. Oh well. Thanks for the awesome build.

**Comment by brett0712** [comment]
I think it's important to remember that Jyers is doing this because he wants to. I supported him when he first posted the Rewrite on Reddit because of how much better his firmware is compared to other offerings. My personal opinion is that his efforts would be best spent on making the firmware the best  it can be instead of constantly compiling builds to suit everyone's specific needs. If this was a paid service, things would be different. 

Compiling your own firmware can certainly seem daunting, but Jyers has done 98% of the work for us by including the various config files. Simply find one that closely aligns with your setup, change the variables you need to, compile, and you're done.

For example, compiling one of the versions Jyers has supplied for a 4.2.7 board only requires the user to enter a "27" then click compile. 

I guess my point is that I'd hate to see this project get derailed because the creator loses interest due to focusing on others specific needs instead of the project.

**Reply by Schnabulation** [reply]
...and don't forget: You can use his firmware and add your very own (basically default) Marlin config-files to it. There are only about 5 lines you need to change to your personal config-file to be able to use it with Jyers firmware.

Also, I'd like to add: if you spend the time and get into "git" it is even simpler: I have my own fork of Jyers firmware with my own config files. When he adds a new revision I have to run 3 commands for my own .bin.

**Reply by bakaneko59** [reply]
Yeah, I really need to figure out how to compile my own. After 35 years in IT you'd think it would be child's play but in all that time I never did learn how to code very well 😁. Ask me to crash a server though and I'm all over it. 😜

**Reply by Schnabulation** [reply]
Ok, now I'm intrigued (I also work in IT): What would your approach be to crash a server?

**Reply by bakaneko59** [reply]
Easiest way is to have your KVM set to server "a" when you meant to be on server "b", then format a drive. Or accidentally delete one of your virtual server drives. Combine these with not verifying your backups and you're in (out) of business. But above all, never let the higher ups know it was something stupid you did. "Hey, computers aren't perfect. It could be a software or hardware problem. I'll run some diagnostics and let you know." and of course conveniently forget to let them know. Or use it as an excuse to get that new hardware/software you've been wanting😜.

**Comment by Jyers** [comment]
Big update for this! I have now coded a GitHub workflow to compile a ton of configurations automatically for each release. This includes a 4.2.7 version for every config :)

**Reply by GalvanicGlaze** [reply]
Wow that's great! Thanks for your efforts!

**Reply by GalvanicGlaze** [reply]
One more suggestion that i have is with that release folder that you have, it might be better to call them 4.2.2 instead of V4 and 4.2.7 instead if v427. Its obvious which is which but i feel like it would still confuse some new people.

**Reply by Jyers** [reply]
I might change this is the future. The only reason they are named that way is because of how the variables worked out in the workflow, but it would be an easy enough modification.
